Mathlib/RingTheory/NonUnitalSubsemiring/Basic.lean

```diff
@@ -0,0 +1,1017 @@
+/-
+Copyright (c) 2022 Jireh Loreaux All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+
+! This file was ported from Lean 3 source module ring_theory.non_unital_subsemiring.basic
+! leanprover-community/mathlib commit 1f0096e6caa61e9c849ec2adbd227e960e9dff58
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Algebra.Ring.Equiv
+import Mathlib.Algebra.Ring.Prod
+import Mathlib.Data.Set.Finite
+import Mathlib.GroupTheory.Submonoid.Membership
+import Mathlib.GroupTheory.Subsemigroup.Membership
+import Mathlib.GroupTheory.Subsemigroup.Centralizer
+
+/-!
+# Bundled non-unital subsemirings
+
+We define bundled non-unital subsemirings and some standard constructions:
+`complete_lattice` structure, `subtype` and `inclusion` ring homomorphisms, non-unital subsemiring
+`map`, `comap` and range (`srange`) of a `non_unital_ring_hom` etc.
+-/
+
+-- porting note: `BigOperators` is open by defaults
+-- open BigOperators
+
+universe u v w
+
+variable {R : Type u} {S : Type v} {T : Type w} [NonUnitalNonAssocSemiring R] (M : Subsemigroup R)
+
+/-- `non_unital_subsemiring_class S R` states that `S` is a type of subsets `s ⊆ R` that
+are both an additive submonoid and also a multiplicative subsemigroup. -/
+class NonUnitalSubsemiringClass (S : Type _) (R : outParam <| Type u) [NonUnitalNonAssocSemiring R]
```

Comment:
Following #1832, we should not have `outParam`s here, since there's already an `outParam` in the parent `SetLike`.
```suggestion
class NonUnitalSubsemiringClass (S : Type _) (R : Type u) [NonUnitalNonAssocSemiring R]
```